hyperspy/_signals/signal2d.py

```diff
@@ -646,28 +646,53 @@ def calibrate(
                 raise ValueError(
                         "With interactive=False x0, y0, x1, y1 and new_length "
                         "must be set.")
+
             self._calibrate(
-                    x0, y0, x1, y1, new_length,
-                    scaled_input=scaled_input, units=units)
+                    x0, y0, x1, y1, new_length, units=units)
 
     def _calibrate(
-            self, x0, y0, x1, y1, new_length, scaled_input=True, units=None):
-        scale = self._get_signal2d_scale(
-                x0, y0, x1, y1, new_length, scaled_input)
+            self, x0, y0, x1, y1, new_length, units=None):
+        scale = self._get_signal2d_scale(x0, y0, x1, y1, new_length)
         sa = self.axes_manager.signal_axes
         sa[0].scale = scale
         sa[1].scale = scale
         if units is not None:
             sa[0].units = units
             sa[1].units = units
 
-    def _get_signal2d_scale(self, x0, y0, x1, y1, length, scaled_input=True):
+    def _get_signal2d_scale(self, x0, y0, x1, y1, length):
+        dtypes = set([type(i) for i in [x0, y0, x1, y1]])
+        if len(dtypes) == 1:
+            dtype = dtypes.pop()
+            if dtype is int:
+                scaled_input = False
+            elif dtype is float:
+                scaled_input = True
+            else:
+                raise TypeError(
```

Comment:
I think that it should be possible to mix integers and floats here (we do it when indexing and slicing after all). You could simply use ``DataAxis._get_index`` to simplify the code.

Reply:
Done!